### DISCUSSION TITLE
Priorities limiting criterions

### DISCUSSION THREAD

**Discussion by Polux49** [neomutt/neomutt#4530] [Development]
## What would you like to be able to do?

Optimize speed when limiting messages, by prioritizing faster patterns. 

#### Example 

Assume that I have 4000 e-mails in my box, 100 of which are less than 1 week old. I'm looking for recent messages containing the keyword 'blah-blah'.  So I call the `limit` function (key `l`) and give it the pattern `~d <1w ~B blah-blah`. 
The resulting selection process is very long, because apparently NeoMutt looks for the keyword `blah-blah` in **all** the 4000 messages, not only in the 100 recent ones.
I wish that, when a limiting pattern combines several sub-patterns like this one, some of which are known to be much longer than others in general, like `~B EXPR`: 
- either NeoMutt treats the slower sub-patterns **after** the others, and **only** to messages matching the faster ones;
-  or it takes care of the order given by the user (so, if the user gives pattern `~d <1w ~B blah-blah`, NeoMutt first limits the index to the 100 messages matching `~d <1w`, and then to the messages **among** those 100 ones matching `~B blah-blah`).

## Have you tried any alternative methods?

With the same example as above, I can follow these steps:
1. Tag recent messages first. 
2. Move them to a temporary box which I create for this task. 
3. Then call the `limit` function in this box. 
4. Do what I want with the matching messages.
5. After that, move back all the messages in the original box.
6. Remove the temporary box.

This is obviously long and tedious. Can it be automized by a macro? I'm not sure, due to the fact that step 4 can be anything, so the whole process requires at least two macros: one for steps 1-3, another one for steps 5-6.

Moreover, even if I know that the `~B EXPR` pattern is very long, I can easily forget that NeoMutt will apply it to all messages even if I precede it by `~d <1w`. 

I would prefer to call the `limit` function without having to think to such workarounds, in order to remain focused on reading messages and looking for information in them.

---
EDIT

Just after posting this message, I discovered the `=B EXPR` pattern of NeoMutt, which does not exist in Mutt. This is a life-changing feature, many thanks for that! :+1:

Well, I still think that my suggestion above is worth being considered, as it improves the efficiency of the `limit` function in any case.

**Comment by flatcap** [comment]
Sorry for not replying sooner...

What you've described -- searching all the emails -- sounds more like a bug.
We'll need to investigate that.

As for the idea to **weight** the search terms, that sounds like a great idea.
It would need to be done carefully, so that the logical meaning isn't altered ([Augustus De Morgan](https://en.wikipedia.org/wiki/Augustus_De_Morgan) may be helpful :-)

Unless we get a volunteer, it'd be low priority, but I'll add it to the todo list.
Thanks.

**Comment by Polux49** [comment]
Thank you! 
:smiley:
